### PR TITLE
PKCE support with oauthlib 3

### PR DESCRIPTION
**Pull request by Maronato**
This PR modifies #656's implementation to use oauthlib's [built-in support for PKCE](https://github.com/oauthlib/oauthlib/pull/617).

Since this version of oauthlib is based on version 3, I had to modify some test cases to comply with the changes.

I also modified the tox and setup files to use the new version of oauthlib.
All modifications to the base test cases and tox/setup files are contained in a491bb1.

### CODE REVIEW

**Comment by auvipy** [comment]
could anyone check why the build is failing?

**Comment by Maronato** [comment]
Seems to be because of my attempt to support oauthlib 3 unofficially.
This dep is to blame:
https://github.com/jazzband/django-oauth-toolkit/blob/a491bb1994ae94e6d0ac9adca632185c805b6ef5/tox.ini#L23

**Comment by Maronato** [comment]
I'll fix it once I get a hold of my pc

**Comment by auvipy** [comment]
ping me after making travis green.

**Comment by auvipy** [comment]
https://github.com/oauthlib/oauthlib/releases seems 3.0.1 is the latest

**Comment by svanscho** [comment]
Any progress? Would love to use this in our new app!

**Comment by sindrig** [comment]
Same here. We're anxiously waiting for this to be able to supply our vendors with the ability to use PKCE.

File: oauth2_provider/migrations/0007_auto_20181019_0349.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# Generated by Django 2.1.1 on 2018-10-19 06:49
```
Comment (auvipy) [inline oauth2_provider/migrations/0007_auto_20181019_0349.py]: can you remove this file and pull from upstream and regen? or could you check the migrations failures? If you don't have time and have any fix in mind let me know, will push on your branch.

**Comment by Maronato** [comment]
@auvipy Sorry it took so long. Fixed the migration and pulled from master.

**Comment by auvipy** [comment]
really great work! thanks for tackling this and reviewers for reviewing.

**Comment by svanscho** [comment]
Thanks guys! Awesome work :-) 👍

File: tests/test_authorization_code.py [hunk 2]
```diff
@@ -537,6 +539,40 @@ def get_auth(self):
...
+            "state": "random_state_string",
+            "scope": "read write",
+            "redirect_uri": "http://example.org",
+            "response_type": "code",
+            "allow": True,
+            "code_challenge": code_challenge,
+            "code_challenge_method": code_challenge_method,
+        }
+
+        response = self.client.post(reverse("oauth2_provider:authorize"), data=authcode_data)
```
Comment (Abhishek8394) [inline tests/test_authorization_code.py]: I had a quick question about this. Shouldn't this be a GET request? I have been trying to get pkce to run against master but couldn't get  it to work. I would appreciate any pointers no how to get it working.

**Comment by n2ygk** [comment]
Yes, GET and optionally POST.
https://tools.ietf.org/html/rfc6749#section-3.1

**Comment by Abhishek8394** [comment]
Here are the two issues I ran into:

1. GET requests fail, both in tests and actual endpoint usage.

2. POST fails due to csrf tokens missing, and I think tests should have
checked for it. Should the tests be checking for this or is it entirely
intentional to skip that?

I think having POST requests will be more secure in that browsers won't
store our code challenges in history, however removing CSRF protection
might introduce bigger issues.

However the spec says GET should work too. So can I open an issue to have
PXCE working on GET requests; and work (get assigned) on it?

**Comment by Maronato** [comment]
This is odd. This pull request does not change anything related to how the requests are handled. It only adds new fields to the existing ones. Would you mind sharing how you are making the requests?

**Comment by Abhishek8394** [comment]
I simply enter the prepared url in browser. Although,  the app is
registered as public app and pkce setting is set to true.

I can provide a more thorough way to recreate it by end of today or
tomorrow. I'll also double check to make sure I'm not messing something on
my side.

**Comment by Maronato** [comment]
I just tested it and you are correct. I was under the impression that oauthlib would capture PKCE credentials, but it doesn't. I've submitted a PR fixing the issue: #707

**Comment by Abhishek8394** [comment]
Awesome! That was quick!

**Pull request merged**